test_conformance/api/test_mem_object_properties_queries.cpp

```diff
@@ -0,0 +1,301 @@
+        {
+            test_object =
+                clCreateImageWithProperties(context, NULL, flags, &format,
+                                            &desc, local_data.data(), &error);
+        }
+        else
+        {
+            test_object = clCreateImageWithProperties(
+                context, test_case.properties.data(), flags, &format, &desc,
+                local_data.data(), &error);
+        }
+        test_error(error, "clCreateImageWithProperties failed");
+    }
+    if (test_case.obj_t == buffer)
+    {
+        if (test_case.properties.size() == 0)
+        {
+            test_object = clCreateBufferWithProperties(
+                context, NULL, flags, local_data.size() * sizeof(cl_uint),
+                local_data.data(), &error);
+        }
+        else
+        {
+            test_object = clCreateBufferWithProperties(
+                context, test_case.properties.data(), flags,
+                local_data.size() * sizeof(cl_uint), local_data.data(), &error);
+        }
+
+        test_error(error, "clCreateBufferWithProperties failed.");
+    }
+    clGetMemObjectInfo(test_object, CL_MEM_PROPERTIES, 0, NULL, &set_size);
+    test_error(error,
+               "clGetMemObjectInfo failed asking for CL_MEM_PROPERTIES.");
+
+    // verify set_size 0 returned
+    if ((test_case.properties.size() == 0
+         || (test_case.properties.size() == 1 && test_case.properties[0] == 0))
+        && set_size == 0)
+    {
+        return TEST_PASS;
+    }
+
+    if (test_case.properties.size() > 1)
+    {
+        if (test_case.properties.size() * sizeof(cl_mem_properties) > set_size)
+        {
+            log_error(
+                "Incorrect properties size returned by clGetMemObjectInfo\n");
+            return TEST_FAIL;
+        }
+    }
+
+    cl_uint number_of_props = set_size / sizeof(cl_mem_properties);
+    object_properties_check.resize(number_of_props);
+    clGetMemObjectInfo(test_object, CL_MEM_PROPERTIES, set_size,
+                       object_properties_check.data(), NULL);
+    test_error(error,
+               "clGetMemObjectInfo failed asking for CL_MEM_PROPERTIES.");
+
+    // check list with 0 terminator is returned
+    if (object_properties_check.size() == 1 && object_properties_check[0] == 0
+        && test_case.properties.size() == 0)
+    {
+        return TEST_PASS;
+    }
+    if (object_properties_check.size() == 1 && object_properties_check[0] == 0
+        && test_case.properties.size() == 1 && test_case.properties[0] == 0)
+    {
+        return TEST_PASS;
+    }
+
+
+    for (auto set_property : test_case.properties)
+    {
+        std::vector<cl_mem_properties>::iterator it =
+            std::find(object_properties_check.begin(),
+                      object_properties_check.end(), set_property);
+        if (it == object_properties_check.end())
+        {
+            log_error("ERROR: Property not found ...");
+            return TEST_FAIL;
+        }
+    }
```

Comment:
This is getting very close - thank you! - but I think it's still not quite right.  In the general case, there are up to three conditions to check:

1. If the passed-in properties (`test_case.properties`) are `NULL` or consist only of the property terminator `0`, and if the queried property size is zero, then the test passes.  This check is on Line 97 - Line 102 and looks correct.

Otherwise, when the queried property size is non-zero:

2. For each passed-in property name-value pair, the same property name-value pair must be in the queried properties (`object_properties_check`) or the test fails.  This check is partially implemented on Line 139, but not completely, because it is looking at the individual properties and values and not property name-value pairs.  To check this condition, we should step through the passed-in properties and the queried properties by two, since we're checking for property name-value pairs.

    a. Note that this check is only needed in the general case, and not in this specific case, since we do not currently pass in any property name-value pairs for this test (yet).

3. The list of queried properties must end with the property terminator `0` or the test fails.  This check is partially implemented on Line 122 - Line 131, but we should check for the property terminator regardless of the size of the passed-in properties and the size of the queried properties.

I can provide the code to implement these checks if it would be helpful, just let me know.  Thanks!

Reply:
Minor update for (1): it's possible that this condition is only valid when the passed-in properties are `NULL`, and not when the passed-in properties consist only of the property terminator `0`.  See this spec question:

https://github.com/KhronosGroup/OpenCL-Docs/issues/317#issuecomment-666740492

Reply:
Please, check now, I hope now all is addressed

Reply:
Similar code will be entered to https://github.com/KhronosGroup/OpenCL-CTS/pull/858

Reply:
Thanks!  I think this has all of the right checks now.

Can you please verify that the checks on Line 102 and Line 126 are necessary?  I think these conditions are already checked by other code paths but I could be mistaken.  If they aren't needed it would be nice to remove them to simplify the test.

Reply:
I agree these conditions are not needed

Reply:
See related comment here: https://github.com/KhronosGroup/OpenCL-CTS/pull/858#discussion_r465470110

Reply:
> See related comment here: [#858 (comment)](https://github.com/KhronosGroup/OpenCL-CTS/pull/858#discussion_r465470110)

fixed